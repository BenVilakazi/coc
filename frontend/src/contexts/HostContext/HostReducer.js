/* eslint-disable no-use-before-define */
function createLobby(state) {
  return {
    ...state,
    gameState: 'waiting-for-players',
  };
}

function getPacks(state) {
  return {
    ...state,
    loading: [...state.loading, 'getting-packs'],
  };
}

function packsReceived(state) {
  return {
    ...state,
    loading: state.loading.filter(
      (loadingVal) => loadingVal !== 'getting-packs',
    ),
  };
}

function clearSubmittedCards(state) {
  const newState = { ...state };
  const { playerIDs, players } = newState;
  playerIDs.forEach((playerID) => {
    players[playerID].submittedCards = [];
  });

  return newState;
}

function dealWhiteCards(state) {
  const {
    deck,
    playerIDs,
    players,
    gameSettings: { handSize },
  } = state;
  const newWhiteCards = [...deck.white];

  // get all active players (not disconnected or waiting to join)
  const activePlayerIDs = playerIDs.filter(
    (playerID) => players[playerID].status === 'playing',
  );

  const neededCardsPerPlayer = activePlayerIDs.map((playerID) => {
    const player = players[playerID];
    return {
      playerID,
      cardsNeeded: handSize - player.cards.length,
    };
  });

  const cardsGivenToPlayers = neededCardsPerPlayer.map(
    ({ playerID, cardsNeeded }) => {
      const newCards = [...players[playerID].cards];
      for (let i = 0; i < cardsNeeded; i += 1) {
        const selection = Math.floor(Math.random() * newWhiteCards.length);
        newCards.push(newWhiteCards.splice(selection, 1)[0]);
      }
      return {
        playerID,
        cards: newCards,
      };
    },
  );

  const newPlayers = cardsGivenToPlayers.reduce((acc, { playerID, cards }) => {
    acc[playerID] = {
      ...players[playerID],
      cards,
    };
    return acc;
  }, {});

  return {
    ...state,
    deck: {
      black: [...deck.black],
      white: newWhiteCards,
    },
    players: { ...players, ...newPlayers }, // keep players that are not currently active
    gameState: 'waiting-to-receive-cards',
    czarSelection: '',
  };
}

function playerReconnected(
  state,
  { playerId: reconnectingPlayerId, playerName },
) {
  const oldPlayerIdIndex = state.playerIDs.findIndex(
    (playerId) => state.players[playerId].name === playerName,
  );

  const oldPlayerId = state.playerIDs[oldPlayerIdIndex];

  const playerStatus = state.players[oldPlayerId].hasSubmittedCards
    ? 'staging'
    : 're-connected';

  const reconnectingPlayerData = {
    ...state.players[oldPlayerId],
    oldIds: [...state.players[oldPlayerId].oldIds, oldPlayerId],
    status: playerStatus,
  };

  const nonReconnectingPlayerIds = state.playerIDs.filter(
    (id) => id !== oldPlayerId,
  );

  const newPlayerIds = [...nonReconnectingPlayerIds];
  newPlayerIds.splice(oldPlayerIdIndex, 0, reconnectingPlayerId);

  const newPlayersData = newPlayerIds.reduce((acc, playerId) => {
    // This is essentially how we are re-assigning the ID of the existing player reconnecting
    if (reconnectingPlayerId === playerId)
      return { ...acc, [playerId]: reconnectingPlayerData };
    return { ...acc, [playerId]: state.players[playerId] };
  }, {});

  // If the czar has selected a card to preview, we need to change this to the new Id to prevent
  // the player from being undefined when the submit button is pressed
  const czarSelection =
    state.czarSelection === oldPlayerId
      ? reconnectingPlayerId
      : state.czarSelection;

  return {
    ...state,
    players: newPlayersData,
    playerIDs: newPlayerIds,
    czarSelection,
  };
}

function playerConnected(state, { playerId, playerName }) {
  // push the new player to the players object, but do not put them in play yet
  const newPlayer = {
    name: playerName,
    score: 0,
    isCzar: false,
    submittedCards: [0],
    hasSubmittedCards: false,
    cards: [],
    status: 'staging',
    oldIds: [],
  };

  return {
    ...state,
    players: {
      ...state.players,
      [playerId]: newPlayer,
    },
    playerIDs: [...state.playerIDs, playerId],
  };
}

function playerSubmitCards(state, { selectedCards, playerId }) {
  const newState = {
    ...state,
    players: {
      ...state.players,
      [playerId]: {
        ...state.players[playerId],
        submittedCards: selectedCards,
        hasSubmittedCards: true,
      },
    },
  };

  const { players, playerIDs } = newState;

  newState.gameState = playerIDs.every(
    (playerID) =>
      players[playerID].isCzar ||
      players[playerID].submittedCards.length ||
      players[playerID].status !== 'playing',
  )
    ? 'czar-select-winner'
    : newState.gameState;

  return newState;
}

function removeSubmittedCards(state) {
  const { players, playerIDs } = state;
  const newPlayers = playerIDs.reduce((acc, playerId) => {
    const player = players[playerId];

    const newCards = player.cards?.filter(
      (card, index) => !player.submittedCards?.includes(index),
    );

    return {
      ...acc,
      [playerId]: {
        ...player,
        cards: newCards,
        hasSubmittedCards: false,
      },
    };
  }, {});

  return {
    ...state,
    players: newPlayers,
  };
}

function disconnectPlayer(state, { playerId }) {
  const removingCzar = state.players[playerId]?.isCzar;

  // add conditional to check if the player is disconnecting due to duplicate name
  if (!state.playerIDs.includes(playerId)) {
    return state;
  }

  const newState = {
    ...state,
    players: {
      ...state.players,
      [playerId]: {
        ...state.players[playerId],
        status: 'disconnected',
      },
    },
  };

  if (
    removingCzar &&
    state.playerIDs.filter((id) =>
      ['playing', 'staging'].includes(state.players[id].status),
    ).length > 1 &&
    !['game-over', 'showing-winning-cards'].includes(state.gameState)
  ) {
    return dealWhiteCards(
      clearSubmittedCards(
        setPlayersToPlaying(
          removeSubmittedCards(setNextCzar(setBlackCard(newState))),
        ),
      ),
    );
  }
  return newState;
}

function kickPlayer(state, { playerId }) {
  const removingCzar = state.players[playerId]?.isCzar;

  // Removes the value playerId from the original playerIDs array
  const newPlayerIds = state.playerIDs.filter(
    (playerID) => playerID !== playerId,
  );

  // Removes the property [playerId] from the original players object
  const { [playerId]: disconnectedPlayer, ...newPlayers } = {
    ...state.players,
  };

  const newState = {
    ...state,
    players: newPlayers,
    playerIDs: newPlayerIds,
  };

  if (
    removingCzar &&
    newPlayerIds.filter((id) =>
      ['playing', 'staging'].includes(state.players[id].status),
    ).length > 1 &&
    !['game-over', 'showing-winning-cards'].includes(state.gameState)
  ) {
    return dealWhiteCards(
      clearSubmittedCards(
        setPlayersToPlaying(
          removeSubmittedCards(setNextCzar(setBlackCard(newState))),
        ),
      ),
    );
  }
  return newState;
}

function kickPlayers(state, { players }) {
  let newState = { ...state };
  players.forEach((playerId) => {
    newState = kickPlayer(newState, { playerId });
  });
  return newState;
}

function czarSelectWinner(state) {
  return {
    ...state,
    gameState: 'selecting-winner',
  };
}

function previewWinner(state, { highlightedPlayerID }) {
  const selectedPlayerId = state.playerIDs.find((playerId) => {
    if (highlightedPlayerID === playerId) return true;
    return state.players[playerId].oldIds.includes(highlightedPlayerID);
  });

  if (state.gameState === 'selecting-winner') {
    return {
      ...state,
      czarSelection: selectedPlayerId,
    };
  }
  return {
    ...state,
  };
}

function selectWinner(state) {
  const roundWinner = state.players[state.czarSelection];

  return {
    ...state,
    gameState: 'showing-winning-cards',
    players: {
      ...state.players,
      [state.czarSelection]: {
        ...roundWinner,
        score: roundWinner.score + 1,
      },
    },
  };
}

function setLobbyId(state, { id }) {
  return {
    ...state,
    lobbyID: id,
  };
}

function startGame(state) {
  // remove dummy submitted cards
  const newPlayers = Object.entries(state.players).reduce((acc, [key, val]) => {
    acc[key] = { ...val };
    acc[key].submittedCards = [];
    return acc;
  }, {});
  return {
    ...state,
    gameState: 'waiting-for-deck',
    players: newPlayers,
  };
}

function setGameSettings(state, { gameSettings }) {
  return {
    ...state,
    gameSettings,
  };
}

function setNextCzar(state) {
  // if there is currently a czar, set the czar to the next connected player in the array
  // else, pick a random czar
  const { players, playerIDs } = state;

  if (playerIDs.length) {
    // find the current czar
    const currentCzar = playerIDs.find((player) => players[player].isCzar);
    // create a list of valid selections for next czar
    const connectedPlayerIDs = playerIDs.filter(
      (id) => players[id].status === 'playing',
    );

    // set the new czar to the old one + 1 in the array, or zero if at the end
    const nextIndex =
      connectedPlayerIDs.indexOf(currentCzar) < connectedPlayerIDs.length - 1
        ? connectedPlayerIDs.indexOf(currentCzar) + 1
        : 0;

    // set the czar to the next one in order, or pick at random
    const newCzar = currentCzar
      ? connectedPlayerIDs[nextIndex]
      : connectedPlayerIDs[
          Math.floor(Math.random() * connectedPlayerIDs.length)
        ];

    // set the new czar in the players object.
    const newPlayers = Object.entries(players).reduce((acc, [key, val]) => {
      acc[key] = { ...val };
      acc[key].isCzar = false;
      return acc;
    }, {});

    return {
      ...state,
      players: {
        ...newPlayers,
        [newCzar]: {
          ...newPlayers[newCzar],
          isCzar: true,
        },
      },
    };
  }

  // If there are no players, return unaltered state
  return { ...state };
}

function setBlackCard(state) {
  const { deck } = state;
  const selectedCard =
    deck.black[Math.floor(Math.random() * state.deck.black.length)];

  const updatedBlackCards = [...state.deck.black];
  updatedBlackCards.splice(updatedBlackCards.indexOf(selectedCard), 1);

  return {
    ...state,
    deck: {
      ...deck,
      black: updatedBlackCards,
    },
    selectedBlackCard: selectedCard,
  };
}

function getDeck(state) {
  return {
    ...state,
    loading: [...state.loading, 'getting-deck'],
  };
}

function setDeck(state, { deck }) {
  return {
    ...state,
    deck,
    loading: state.loading.filter(
      (loadingVal) => loadingVal !== 'getting-deck',
    ),
  };
}

function getJoinCode(state) {
  return {
    ...state,
    loading: [...state.loading, 'join-code'],
  };
}

function updateJoinCode(state, { lobbyID }) {
  return {
    ...state,
    lobbyID,
    loading: state.loading.filter((loadingVal) => loadingVal !== 'join-code'),
  };
}

function setPlayersToPlaying(state) {
  const newState = { ...state };

  newState.playerIDs.forEach((player) => {
    const playerStatus = newState.players[player].status;
    if (playerStatus === 'staging' || playerStatus === 're-connected') {
      newState.players[player].status = 'playing';
    }
  });

  return {
    ...newState,
  };
}

function toggleJoinCode(state) {
  return {
    ...state,
    gameSettings: {
      ...state.gameSettings,
      hideJoinCode: !state.gameSettings.hideJoinCode,
    },
  };
}

function gameOver(state, { gameWinner }) {
  return {
    ...state,
    gameState: 'game-over',
    gameWinner,
  };
}

function HostReducer(state, action) {
  const { type, payload } = action;
  switch (type) {
    case 'CREATE_LOBBY':
      return createLobby(state);

    case 'GET_PACKS':
      return getPacks(state);

    case 'PACKS_RECEIVED':
      return packsReceived(state);

    case 'PLAYER_CONNECTED':
      return playerConnected(state, payload);

    case 'PLAYER_RECONNECTED':
      return playerReconnected(state, payload);

    case 'PLAYER_DISCONNECTED':
      return disconnectPlayer(state, payload);

    case 'PLAYER_SUBMIT':
      return playerSubmitCards(state, payload);

    case 'KICK_PLAYER':
      return kickPlayer(state, payload);

    case 'SKIP_UNSUBMITTED_PLAYERS':
    case 'CZAR_SELECT_WINNER':
      return czarSelectWinner(state);

    case 'PREVIEW_WINNER':
      return previewWinner(state, payload);

    case 'WINNER_SELECTED':
      return selectWinner(state);

    case 'SET_LOBBY_ID':
      return setLobbyId(state, payload);

    case 'START_GAME':
      return startGame(state);

    case 'SET_GAME_SETTINGS':
      return setGameSettings(state, payload);

    case 'SET_NEXT_CZAR':
      return setNextCzar(state);

    case 'SELECT_BLACK_CARD':
      return setBlackCard(state);

    case 'GET_DECK':
      return getDeck(state);

    case 'SET_DECK':
      return setDeck(state, payload);

    case 'DEAL_WHITE_CARDS':
      return dealWhiteCards(clearSubmittedCards(state));

    case 'SEND_CARDS_TO_PLAYERS':
      return setPlayersToPlaying(state);

    case 'REMOVE_SUBMITTED_CARDS_FROM_PLAYER':
      return removeSubmittedCards(state);

    case 'SHUFFLE_JOIN_CODE':
      return getJoinCode(state);

    case 'UPDATE_JOIN_CODE':
      return updateJoinCode(state, payload);

    case 'ADD_PLAYERS_FROM_STAGING':
      return setPlayersToPlaying(state);

    case 'TOGGLE_JOIN_CODE_VISIBILITY':
      return toggleJoinCode(state);

    case 'TOO_MANY_PLAYERS':
      return kickPlayers(state, payload);

    case 'GAME_OVER':
      return gameOver(state, payload);

    default:
      return { ...state };
  }
}

export default HostReducer;
