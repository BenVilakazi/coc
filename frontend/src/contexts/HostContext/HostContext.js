import React, { createContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import useReducerMiddleware from '../useReducerMiddleware';
import HostReducer from './HostReducer';
import socketInstance from '../../socket/socket';
import hostReducerMiddleware from './hostReducerMiddleware';
import config from '../../config';
import AlertModal from '../../components/Modal/AlertModal';

const propTypes = {
  children: PropTypes.node.isRequired,
};

const { emitter } = socketInstance;

const initialState = {
  gameState: 'waiting-for-lobby',
  lobbyID: '',
  players: {},
  playerIDs: [],
  gameSettings: {
    maxPlayers: config.maxPlayers.default,
    winningScore: config.winningScore.default,
    selectedPacks: config.initialSelectedPack,
    handSize: config.handSize,
    winnerScreenDisplayTime: config.winnerScreenDisplayTime,
    hideJoinCode: false,
  },
  deck: { black: [], white: [] },
  loading: [],
  czarSelection: '',
};

export const HostContext = createContext();

function HostProvider({ children }) {
  const [state, dispatch] = useReducerMiddleware(
    hostReducerMiddleware,
    HostReducer,
    initialState,
  );

  const [socketHasError, setSocketHasError] = useState(false);

  function createMessageHandler(currentState) {
    const czarId =
      currentState.playerIDs.length > 0 &&
      currentState.playerIDs.find(
        (playerId) => currentState.players[playerId].isCzar,
      );

    return ({ event, payload, sender }) => {
      switch (event) {
        case 'player-connected': {
          const allPlayers = [...Object.values(currentState.players)];
          const existingPlayer = allPlayers.find(
            (player) =>
              player.name.toUpperCase() === payload.playerName.toUpperCase(),
          );

          if (!existingPlayer) {
            return dispatch({ type: 'PLAYER_CONNECTED', payload });
          }

          if (existingPlayer.status === 'disconnected') {
            return dispatch({ type: 'PLAYER_RECONNECTED', payload });
          }

          return dispatch({ type: 'SEND_NAME_TAKEN_MESSAGE', payload });
        }
        case 'player-disconnected':
          return dispatch({
            type: 'PLAYER_DISCONNECTED',
            payload: {
              ...payload,
              czarId,
            },
          });

        case 'preview-winner':
          return dispatch({ type: 'PREVIEW_WINNER', payload });

        case 'winner-selected':
          return dispatch({ type: 'WINNER_SELECTED', payload });

        case 'lobby-created':
          return dispatch({ type: 'SET_LOBBY_ID', payload });

        case 'black-card-selected':
          return dispatch({ type: 'SELECT_BLACK_CARD', payload });

        case 'join-code-shuffled':
          return dispatch({ type: 'UPDATE_JOIN_CODE', payload });

        case 'player-submit':
          return dispatch({
            type: 'PLAYER_SUBMIT',
            payload: { ...payload, playerId: sender },
          });

        case 'socket-connection-error':
          dispatch({ type: 'UPDATE_JOIN_CODE', payload: { lobbyID: 'ERROR' } });
          return setSocketHasError(true);

        default:
          return undefined;
      }
    };
  }

  useEffect(() => {
    const handleMessage = createMessageHandler(state);
    emitter.on('message', handleMessage);
    return () => {
      emitter.off('message', handleMessage);
    };
  }, [state]);

  useEffect(() => {
    const stagedPlayerIDs = state.playerIDs.filter(
      (playerID) => state.players[playerID].status === 'staging',
    );

    const reconnectingPlayersIds = state.playerIDs.filter(
      (playerID) => state.players[playerID].status === 're-connected',
    );

    if (reconnectingPlayersIds.length > 0) {
      const reConnectingPlayers = reconnectingPlayersIds.reduce(
        (acc, id) => ({ ...acc, [id]: state.players[id] }),
        {},
      );

      dispatch({
        type: 'SEND_CARDS_TO_PLAYERS',
        payload: {
          selectedBlackCard: state.selectedBlackCard,
          players: reConnectingPlayers,
          playerIDs: reconnectingPlayersIds,
        },
      });
    }

    if (stagedPlayerIDs.length) {
      const numberOfPlayersExceedingLimit =
        state.playerIDs.length - state.gameSettings.maxPlayers;

      if (numberOfPlayersExceedingLimit > 0) {
        const playerIDsExceedingLimit = state.playerIDs.slice(
          -numberOfPlayersExceedingLimit,
        );

        dispatch({
          type: 'TOO_MANY_PLAYERS',
          payload: {
            players: playerIDsExceedingLimit,
          },
        });
      } else {
        const message =
          state.gameState === 'waiting-for-players'
            ? {
                big: "You've joined the lobby",
                small: 'Please wait for the host to start the game',
              }
            : {
                big: 'A round is in progress',
                small: 'You will join the next round automatically',
              };

        dispatch({
          type: 'SEND_PLAYER_CONNECTED_MESSAGES',
          payload: {
            players: stagedPlayerIDs,
            message,
          },
        });
      }
    }
  }, [state.playerIDs]);

  useEffect(() => {
    // if there is one player AND they are not waiting for the game to begin, end the game
    if (
      state.playerIDs.filter(
        (playerID) => state.players[playerID].status === 'playing',
      ).length === 1 &&
      state.gameState !== 'game-over'
    ) {
      dispatch({
        type: 'GAME_OVER',
        payload: { gameWinner: state.playerIDs[0], playerIDs: state.playerIDs },
      });
    }
  }, [state.playerIDs, state.players]);

  return (
    <HostContext.Provider value={{ state, dispatch }}>
      {children}
      {socketHasError && (
        <AlertModal
          bigText="SOCKET ERROR"
          smallText="Please try again later"
          buttonText="Click anywhere to restart"
          onClick={() => window.location.reload()}
        />
      )}
    </HostContext.Provider>
  );
}

HostProvider.propTypes = propTypes;

export default HostProvider;
