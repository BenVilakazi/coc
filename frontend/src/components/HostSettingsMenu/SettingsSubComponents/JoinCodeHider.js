import React, { useContext } from 'react';
import PropTypes from 'prop-types';

import OptionButton from './OptionButton';
import { HostContext } from '../../../contexts/HostContext/HostContext';

const propTypes = {
  isEnabled: PropTypes.bool.isRequired,
  onDisabledClick: PropTypes.func.isRequired,
};

function JoinCodeHider({ isEnabled, onDisabledClick }) {
  const { state, dispatch } = useContext(HostContext);
  const hidden = state.gameSettings.hideJoinCode;

  return (
    <OptionButton
      isEnabled={isEnabled}
      onEnabledClick={() => dispatch({ type: 'TOGGLE_JOIN_CODE_VISIBILITY' })}
      onDisabledClick={onDisabledClick}
    >
      {hidden ? 'SHOW JOIN CODE' : 'HIDE JOIN CODE'}
    </OptionButton>
  );
}

JoinCodeHider.propTypes = propTypes;

export default JoinCodeHider;
