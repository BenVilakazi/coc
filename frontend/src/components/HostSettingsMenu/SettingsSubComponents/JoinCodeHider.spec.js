import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import JoinCodeHider from './JoinCodeHider';
import { HostContext } from '../../../contexts/HostContext/HostContext';

// mock the OptionButton component that is imported into JoinCodeHider
jest.mock(
  './OptionButton', // eslint-disable-next-line react/prop-types
  () => ({ isEnabled, onEnabledClick, onDisabledClick, children }) => {
    function clickHandler() {
      if (isEnabled) {
        onEnabledClick();
      } else {
        onDisabledClick();
      }
    }

    return (
      <button
        type="button"
        data-testid="option-button"
        data-is-enabled={isEnabled}
        onClick={clickHandler}
      >
        {children}
      </button>
    );
  },
);

describe('JoinCodeHider', () => {
  describe('rendering', () => {
    it('matches the snapshot', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      const tree = renderer
        .create(
          <HostContext.Provider value={{ state, dispatch }}>
            <JoinCodeHider isEnabled onDisabledClick={() => {}} />
          </HostContext.Provider>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('says "HIDE JOIN CODE" when the hideJoinCode setting is false', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled onDisabledClick={() => {}} />
        </HostContext.Provider>,
      );
      const optionButton = screen.getByTestId('option-button');

      expect(optionButton.textContent).toBe('HIDE JOIN CODE');
    });

    it('says "SHOW JOIN CODE" when the hideJoinCode setting is true', () => {
      const state = {
        gameSettings: {
          hideJoinCode: true,
        },
      };

      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled onDisabledClick={() => {}} />
        </HostContext.Provider>,
      );
      const optionButton = screen.getByTestId('option-button');

      expect(optionButton.textContent).toBe('SHOW JOIN CODE');
    });
  });

  describe('functionality', () => {
    it('passes the expected text to be rendered as text in the option button', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled onDisabledClick={() => {}} />
        </HostContext.Provider>,
      );
      const optionButton = screen.getByTestId('option-button');

      expect(optionButton.textContent).toBe('HIDE JOIN CODE');
    });

    it('passes the isEnabled prop through when false', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled={false} onDisabledClick={() => {}} />
        </HostContext.Provider>,
      );
      const optionButton = screen.getByRole('button', {
        name: 'HIDE JOIN CODE',
      });

      expect(optionButton.dataset.isEnabled).toBe('false');
    });

    it('passes the isEnabled prop through when true', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled onDisabledClick={() => {}} />
        </HostContext.Provider>,
      );
      const optionButton = screen.getByRole('button', {
        name: 'HIDE JOIN CODE',
      });

      expect(optionButton.dataset.isEnabled).toBe('true');
    });

    // this test and the one below it are temporary and to be replaced once functionality
    // for JoinCodeHider is actually built out
    it('does not dispatch an action when the button is not clicked', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled onDisabledClick={() => {}} />
        </HostContext.Provider>,
      );

      expect(dispatch).not.toHaveBeenCalled();
    });

    it('dispatches a "TOGGLE_JOIN_CODE_VISIBILITY" action when button is enabled and clicked', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled onDisabledClick={() => {}} />
        </HostContext.Provider>,
      );
      userEvent.click(screen.getByRole('button', { name: 'HIDE JOIN CODE' }));

      expect(dispatch).toHaveBeenCalledWith({
        type: 'TOGGLE_JOIN_CODE_VISIBILITY',
      });
    });

    it('does not call the onDisabledClick callback when disabled but not clicked', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      const mockOnDisabledClick = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider
            isEnabled={false}
            onDisabledClick={mockOnDisabledClick}
          />
          ,
        </HostContext.Provider>,
      );

      expect(mockOnDisabledClick).not.toHaveBeenCalled();
    });

    it('calls the onDisabledClick callback when disabled and clicked', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      const mockOnDisabledClick = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider
            isEnabled={false}
            onDisabledClick={mockOnDisabledClick}
          />
          ,
        </HostContext.Provider>,
      );
      userEvent.click(screen.getByRole('button', { name: 'HIDE JOIN CODE' }));

      expect(mockOnDisabledClick).toHaveBeenCalled();
    });
  });

  describe('propTypes', () => {
    it('logs an error when attempting to render without the isEnabled Prop', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider onDisabledClick={() => {}} />
        </HostContext.Provider>,
      );
      userEvent.click(screen.getByTestId('option-button'));

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the isEnabled Prop as a non bool', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled="foo" onDisabledClick={() => {}} />
        </HostContext.Provider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render without the onDisabledClick Prop', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled />
        </HostContext.Provider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the onDisabledClick Prop as a non function', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled onDisabledClick="foo" />
        </HostContext.Provider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('does not log an error when all required props are given correctly', () => {
      const state = {
        gameSettings: {
          hideJoinCode: false,
        },
      };

      const dispatch = jest.fn();

      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <JoinCodeHider isEnabled onDisabledClick={() => {}} />
        </HostContext.Provider>,
      );

      expect(consoleSpy).not.toHaveBeenCalled();
    });
  });
});
