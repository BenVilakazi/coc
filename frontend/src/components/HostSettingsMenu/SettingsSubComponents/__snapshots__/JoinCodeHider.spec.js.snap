// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinCodeHider rendering matches the snapshot 1`] = `
<button
  data-is-enabled={true}
  data-testid="option-button"
  onClick={[Function]}
  type="button"
>
  HIDE JOIN CODE
</button>
`;
